docker/archive/dest.go

```diff
@@ -70,3 +75,61 @@ func (d *archiveImageDestination) Close() error {
 func (d *archiveImageDestination) Commit(ctx context.Context, unparsedToplevel types.UnparsedImage) error {
 	return d.Destination.Commit(ctx)
 }
+
+func (m multiImageDestinationReference) NewImageDestination(_ context.Context, _ *types.SystemContext) (types.ImageDestination, error) {
+	return m.dest, nil
+}
+
+// MultiImageDestinations allows for creating and writing to docker archives
+// that include more than one image.
+type MultiImageDestination struct {
+	*archiveImageDestination
+	path string
+}
+
+// multiImageDestinationReference is a types.ImageReference embedding a MultiImageDestination.
+type multiImageDestinationReference struct {
+	*archiveReference
+	dest *MultiImageDestination
+}
+
+// NewMultiImageDestination returns a MultiImageDestination for the specified path.
+func NewMultiImageDestination(sys *types.SystemContext, path string) (*MultiImageDestination, error) {
+	dest, err := newArchiveImageDestination(sys, path, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &MultiImageDestination{dest, path}, nil
+}
+
+// Reference returns an ImageReference embedding the MultiImageDestination.
+func (m *MultiImageDestination) Reference() types.ImageReference {
+	ref := &archiveReference{path: m.path}
+	return &multiImageDestinationReference{ref, m}
+}
+
+// Close is a NOP.  Please use Finalize() for committing the archive and
+// closing the underlying resources.
+func (m *MultiImageDestination) Close() error {
```

Comment:
`Close()` is conceptually a bit different from `Finalize()` (or `ImageDestination.Commit`); it allows cleaning up the temporary files even on error.

Sure, this is a possible way to structure the API, but it’s a bit inconvenient to use: A typically caller will use something like `defer multiDest.Close()` and might not even check for errors if there is already a failure with a more important root cause to preserve, whereas on success the caller really wants to check that `multiDest.Finalize()` did succeed. Having “commit” and “deallocate” be the same operation forces every such caller to have a `committed` flag that’s checked inside the `defer`, or to have a critical part of the process in a `defer`.